pkg/kubelet/kuberuntime/security_context_windows.go

```diff
@@ -0,0 +1,65 @@
+// +build windows
+
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package kuberuntime
+
+import (
+	"fmt"
+	"k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/securitycontext"
+)
+
+var (
+	windowsRootUserName = "ContainerAdministrator"
+)
+
+// verifyRunAsNonRoot verifies RunAsNonRoot on windows.
+// https://github.com/kubernetes/enhancements/blob/master/keps/sig-windows/20190103-windows-node-support.md#v1container
+// Windows does not have a root user, we cannot judge the root identity of the windows container by the way to judge the root(uid=0) of the linux container.
+// According to the discussion of sig-windows, at present, we assume that ContainerAdministrator is the windows container root user,
+// and then optimize this logic according to the best time.
+// https://docs.google.com/document/d/1Tjxzjjuy4SQsFSUVXZbvqVb64hjNAG5CQX8bK7Yda9w
+func verifyRunAsNonRoot(pod *v1.Pod, container *v1.Container, uid *int64, username string) error {
+	effectiveSc := securitycontext.DetermineEffectiveSecurityContext(pod, container)
+	// If the option is not set, or if running as root is allowed, return nil.
+	if effectiveSc == nil || effectiveSc.RunAsNonRoot == nil || !*effectiveSc.RunAsNonRoot {
+		return nil
+	}
+	if effectiveSc.RunAsUser != nil {
+		klog.Warningf("Windows container does not support SecurityContext.RunAsUser, please use SecurityContext.WindowsOptions")
+	}
+	if effectiveSc.SELinuxOptions != nil {
+		klog.Warningf("Windows container does not support SecurityContext.SELinuxOptions, please use SecurityContext.WindowsOptions")
```

Comment:
This is a big deal, right?  Someone thinks they are running with security options that we aren't applying.  Should this fail more loudly?

If i'm interpreting this correctly, I would suggest emiting an event, or even returning an error here.  Same comment for other unsupported options.

Reply:
I think the argumentation here is that these settings were likely set by some plugin. See https://github.com/kubernetes/kubernetes/pull/93475#discussion_r486666465

This is why warning was good enough from my perspective, but I'd like somebody from windows sig to confirm

Reply:
Maybe we should add a check to https://github.com/kubernetes/kubernetes/blob/360659102c8ae94a5bee6e9cbbc3d1160db58428/pkg/apis/core/validation/validation.go#L3769 that verifies only only of SecurityContext.SELinuxOptions or SecurityContext.WindowsOptions are set.

From my perspective we should encourage plugin/automation authors to use node selectors if they are going to blanket apply security policy to pods rather than working around that assumption in the kubelet.
@jsturtevant thoughts?

Reply:
Because these do not cause cri errors, it is just that the features of linux container on windows container cannot take effect, and vice versa. If kubelet fails more loudly, it will affect existing applications (adding wrong WindowsOptions or LinuxOptions). So I think the warning level log may be a way to make sense (may need to be more detailed)

Reply:
@marosset 

I think this is a very good idea. When apiserver pre-verification, if the user explicitly sets `SecurityContext.SELinuxOptions` or `SecurityContext.WindowsOptions` for verification, if the discussion passes, I can try to open another A pr. 
But kubelet is really responsible for calling cri to create pod logic. Should it also have its own necessary verification logic?

Reply:
> Maybe we should add a check to https://github.com/kubernetes/kubernetes/blob/360659102c8ae94a5bee6e9cbbc3d1160db58428/pkg/apis/core/validation/validation.go#L3769 that verifies only one of SecurityContext.SELinuxOptions or SecurityContext.WindowsOptions are set.

Tightening API validation is not generally possible (see https://github.com/kubernetes/kubernetes/issues/64841). If that combination of settings is never legitimate (is a multi-OS image possible?), we could add that to the list of things we want to warn about (https://github.com/kubernetes/kubernetes/issues/94626)

Reply:
Thanks @liggitt, that addresses my concern.

Reply:
Thanks @liggitt, that addresses my concern.

Reply:
This also addresses my concern.